crates/bevy_ecs/src/system/query.rs

```diff
@@ -380,29 +382,37 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Query<'w, 's, D, F> {
     #[inline]
     pub(crate) unsafe fn new(
         world: UnsafeWorldCell<'w>,
-        state: &'s QueryState<D, F>,
+        state: Entity,
         last_run: Tick,
         this_run: Tick,
     ) -> Self {
-        state.validate_world(world.id());
+        let query_state = world.world_mut().get::<QueryState<D, F>>(state).unwrap();
+        query_state.validate_world(world.id());
 
         Self {
             world,
             state,
             last_run,
             this_run,
+            marker: PhantomData,
         }
     }
 
+    fn get_query_state(&self) -> &QueryState<D, F> {
+        return unsafe { self.world.world_mut().get::<QueryState<D, F>>(self.state).unwrap() }
+    }
+
     /// Returns another `Query` from this that fetches the read-only version of the query items.
     ///
     /// For example, `Query<(&mut D1, &D2, &mut D3), With<F>>` will become `Query<(&D1, &D2, &D3), With<F>>`.
     /// This can be useful when working around the borrow checker,
     /// or reusing functionality between systems via functions that accept query types.
     pub fn to_readonly(&self) -> Query<'_, 's, D::ReadOnly, F> {
-        let new_state = self.state.as_readonly();
+        let mut query_state = self.get_query_state();
+        let new_state = query_state.as_readonly().clone();
+        let read_only_state = unsafe { self.world.world_mut() }.spawn(*new_state).id();
```

Comment:
We don't need another entity here. `as_readonly` calls `pointer::cast` and so doesn't need to be copied.

I am not sure the ideal solution but I think type erasing `QueryState<D, F>` and "transmuting" that type erased form into a `QueryState<D, F>` should work here as the underlying bytes are the same between the two forms.

However another reason to call `clone` here is ownership, specifically how do we share ownership of the original `QueryState<D, F>` since hypothetically both `Query` should hold a immutable reference to the `QueryState<D, F>` if it is shared like that. (Of course one would treat it as a `QueryState<D::ReadOnly, F>` but that is a pointer cast thing)

Reply:
With how `Query` is defined right now you need another entity, because the `Query<D::ReadOnly, F>` will try to access a component of type `QueryState<D::ReadOnly, F>` on that `Entity`, but that may only have a component of type `QueryState<D, F>`.

`Query` would have to be changed to store a `&'w QueryState<D, F>` to do what you're proposing.

Reply:
That is why I said it needs to be type erased, aka store the `QueryState` in a way that isn't type specific.